### PR TITLE
Add EntityType property to IEdmNavigationSource

### LINKED ISSUE #2820: Consider adding an `EntityType` property to `IEdmNavigationSource`
This issue mentions the perf overhead of the `IEdmNavigationSource.EntityType()` extension method. I made slight improvements to that method but I think it would be easier if `IEdmNavigationSource` had an `EntityType` property that could be easily accessed without casting. Each implementation of `IEdmNavigationSource` could store the entity type in the property so that it can be accessed without having to recompute it every time (for example calling `AsElementType()` for entity sets).

Generally, I think we should revisit other EDM interfaces where we are doing recurring computations and casting that could be simplified and made more performant with simpler/unified interfaces with the right properties, e.g. see: https://github.com/OData/odata.net/issues/2528

### PR DESCRIPTION
**Pull request by gathogojr**
### Issues

*This pull request fixes #2820.*

### Description

This pull request adds a `EntityType` property to `IEdmNavigationSource` interface to eliminate the constant type casting that happens when `EntityType(IEdmNavigationSource)` extension method is called.

The extension method is invoked very frequently and on hot paths in some cases.

Changes included implementing the property in classes that implement the `IEdmNavigationSource` interface.
In the crucial `EdmNavigationSource` class, the implementation looks as follows:
```csharp
// ...

private IEdmEntityType entityType;
private bool entityTypeSet = false;

// ...

public IEdmEntityType EntityType
{
    get
    {
        if (!this.entityTypeSet)
        {
            this.entityType = this.Type?.AsElementType() as IEdmEntityType;
            this.entityTypeSet = true;
        }

        return this.entityType;
    }
}
```

Since the `EntityType` extension method is public, I retained it but applied the `Obsolete` attribute.

Refactoring was also done to replace calls to the `EntityType` extension method with `IEdmNavigationSource.EntityType` property access.

**NOTE:**
The build pipeline for 8.x is not running the E2E tests currently.
A previous 8.x pull request got rid of `CreateODataDeltaReader` since it's deprecated and inadvertently forgot to update the tests in _/test/EndToEndTests/Tests/Client/Build.Desktop/DeltaTests/DeltaTests.cs_.
The E2E solution couldn't build locally so I fixed the delta tests in this pull request.
In addition, the tests had very few assert statements with little to no verification of the expected payload structure and data so I refactored them extensively.

### Checklist (Uncheck if it is not completed)

- [x] *Test cases added*
- [x] *Build and test with one-click build and test script passed*

### CODE REVIEW

File: src/Microsoft.OData.Core/UnknownEntitySet.cs [hunk 1]
```diff
@@ -44,6 +45,19 @@ public IEdmType Type
             }
         }
 
+        public IEdmEntityType EntityType
+        {
+            get
+            {
+                if (this.entityType == null)
+                {
+                    this.entityType = this.Type?.AsElementType() as IEdmEntityType;
```
Comment (wachugamaina) [inline src/Microsoft.OData.Core/UnknownEntitySet.cs:54]: Is there a possibility that this casting may result in null ?

File: src/Microsoft.OData.Edm/Csdl/Semantics/CsdlSemanticsNavigationSource.cs [hunk 2]
```diff
@@ -87,6 +90,20 @@ public IEnumerable<IEdmNavigationPropertyBinding> NavigationPropertyBindings
             get { return this.navigationTargetsCache.GetValue(this, ComputeNavigationTargetsFunc, null); }
         }
 
+        public IEdmEntityType EntityType
+        {
+            get
+            {
+                if (!this.entityTypeSet)
```
Comment (wachugamaina) [inline src/Microsoft.OData.Edm/Csdl/Semantics/CsdlSemanticsNavigationSource.cs:97]: Why not check for !this.entityType  == null ?

File: test/EndToEndTests/Tests/Client/Build.Desktop/DeltaTests/DeltaTests.cs [hunk 3]
```diff
@@ -48,20 +48,21 @@ public void RequestDeltaLink()
                 {
                     using (var messageReader = new ODataMessageReader(responseMessage, readerSettings, Model))
                     {
-                        var deltaReader = messageReader.CreateODataDeltaReader(customersSet, customerType);
```
Comment (wachugamaina) [inline test/EndToEndTests/Tests/Client/Build.Desktop/DeltaTests/DeltaTests.cs:51]: This change seems unrelated to this Pr, is it ?

File: test/EndToEndTests/Tests/Client/Build.Desktop/DeltaTests/DeltaTests.cs [hunk 4]
```diff
@@ -161,42 +162,47 @@ public void RequestDeltaLink_Expanded()
...
-                                    case ODataReaderState.Completed:
-                                        Assert.NotNull(deltaReader.Item);
-                                        break;
-                                }
+
+                            }
+                            else if (deltaReader.State == ODataReaderState.NestedResourceInfoEnd)
+                            {
+                                Assert.NotNull(deltaReader.Item as ODataNestedResourceInfo);
+                                //switch (deltaReader.SubState)
```
Comment (wachugamaina) [inline test/EndToEndTests/Tests/Client/Build.Desktop/DeltaTests/DeltaTests.cs:184]: Please delete unused code, you can always retrieve them from Git.

File: src/Microsoft.OData.Core/UnknownEntitySet.cs (hunk 1 above)
Comment (gathogojr) [reply src/Microsoft.OData.Core/UnknownEntitySet.cs:54]: @wachugamaina That's correct. It may result into null and even the previous extension method that was being used in place of this property could return null in some cases. The caller is supposed to anticipate that. I'll mention that it can be null in the property doc.

File: src/Microsoft.OData.Edm/Csdl/Semantics/CsdlSemanticsNavigationSource.cs (hunk 2 above)
Comment (gathogojr) [reply src/Microsoft.OData.Edm/Csdl/Semantics/CsdlSemanticsNavigationSource.cs:97]: @wachugamaina Because the value can be null, I didn't want the expression `this.Type?.AsElementType() as IEdmEntityType;` to be executed multiple times since this property is referenced a lot across the library including in hot paths

File: test/EndToEndTests/Tests/Client/Build.Desktop/DeltaTests/DeltaTests.cs (hunk 3 above)
Comment (gathogojr) [reply test/EndToEndTests/Tests/Client/Build.Desktop/DeltaTests/DeltaTests.cs:51]: Yes it is. The build pipeline for 8.x is not running the E2E tests currently so a previous 8.x pull request got rid of `CreateODataDeltaReader` since it's deprecated and inadvertently forgot to update this test. I tried to build the E2E solution locally and it couldn't run so I fixed it in this pull request

File: test/EndToEndTests/Tests/Client/Build.Desktop/DeltaTests/DeltaTests.cs (hunk 4 above)
Comment (gathogojr) [reply test/EndToEndTests/Tests/Client/Build.Desktop/DeltaTests/DeltaTests.cs:184]: @wachugamaina I'll replace the commented lines with better assert statements. I noticed that tests have insufficient assert statements. Checking that the item is not null doesn't feel sufficient enough to me

**Review by habbes — APPROVED** [review]

File: src/Microsoft.OData.Edm/Schema/AmbiguousEntitySetBinding.cs [hunk 5]
```diff
@@ -41,6 +43,19 @@ public IEdmType Type
             get { return new EdmCollectionType(new EdmEntityTypeReference(new BadEntityType(String.Empty, this.Errors), false)); }
         }
 
+        public IEdmEntityType EntityType
```
Comment (ElizabethOkerio) [inline src/Microsoft.OData.Edm/Schema/AmbiguousEntitySetBinding.cs:46]: Can this method be shared instead of duplicating it in most of these classes?

File: src/Microsoft.OData.Edm/Schema/BadEntitySet.cs [hunk 6]
```diff
@@ -56,6 +57,19 @@ public IEdmType Type
             get { return new EdmCollectionType(new EdmEntityTypeReference(new BadEntityType(String.Empty, this.Errors), false)); }
         }
```
Comment (ElizabethOkerio) [inline src/Microsoft.OData.Edm/Schema/BadEntitySet.cs:59]: Add doc string

File: src/Microsoft.OData.Edm/Schema/AmbiguousSingletonBinding.cs [hunk 7]
```diff
@@ -22,6 +24,19 @@ public IEdmType Type
             get { return new BadEntityType(String.Empty, this.Errors); }
         }
 
+        public IEdmEntityType EntityType
```
Comment (ElizabethOkerio) [inline src/Microsoft.OData.Edm/Schema/AmbiguousSingletonBinding.cs:27]: Add doc string

File: src/Microsoft.OData.Edm/Schema/AmbiguousEntitySetBinding.cs (hunk 5 above)
Comment (ElizabethOkerio) [reply src/Microsoft.OData.Edm/Schema/AmbiguousEntitySetBinding.cs:46]: Add doc string

File: test/EndToEndTests/Tests/Client/Build.Desktop/DeltaTests/DeltaTests.cs [hunk 8]
```diff
@@ -22,217 +25,918 @@ public DeltaTests()
...
-            ODataMessageReaderSettings readerSettings = new ODataMessageReaderSettings() { BaseUri = ServiceBaseUri };
+            var readerSettings = new ODataMessageReaderSettings { BaseUri = ServiceBaseUri };
 
-            foreach (var mimeType in deltaMimeTypes)
-            {
-                var requestMessage = new HttpWebRequestMessage(new Uri(ServiceBaseUri.AbsoluteUri + "$delta", UriKind.Absolute));
-                requestMessage.SetHeader("Accept", mimeType);
-                var responseMessage = requestMessage.GetResponse();
-                Assert.Equal(200, responseMessage.StatusCode);
+            var requestMessage = new HttpWebRequestMessage(new Uri(ServiceBaseUri.AbsoluteUri + "$delta", UriKind.Absolute));
```
Comment (ElizabethOkerio) [inline test/EndToEndTests/Tests/Client/Build.Desktop/DeltaTests/DeltaTests.cs:44]: Is this class from OData Client.. I thought we moved away from it?

File: src/Microsoft.OData.Edm/Schema/AmbiguousEntitySetBinding.cs (hunk 5 above)
Comment (gathogojr) [reply src/Microsoft.OData.Edm/Schema/AmbiguousEntitySetBinding.cs:46]: @ElizabethOkerio Would you be happy with this:
```csharp
        public IEdmEntityType EntityType
        {
            get
            {
                if (this.entityType == null)
                {
#pragma warning disable CS0618 // Type or member is obsolete
                    this.entityType = this.EntityType();
#pragma warning restore CS0618 // Type or member is obsolete
                }

                return this.entityType;
            }
        }
```
Basically mark the existing `EntityType` extension method as `Obsolete` with the message that it'll not be publicly accessible starting 9.0. But because it's obsolete and I'm calling it from this property (to avoid repeating `this.Type.AsElementType() as IEdmEntityType;`  in a lot of places), apply the `#pragma warning disable` statements to suppress the suggestion/warning by Visual Studio to mark to property as obsolete too...

File: test/EndToEndTests/Tests/Client/Build.Desktop/DeltaTests/DeltaTests.cs (hunk 8 above)
Comment (gathogojr) [reply test/EndToEndTests/Tests/Client/Build.Desktop/DeltaTests/DeltaTests.cs:44]: @ElizabethOkerio This class is not from the OData client library. It is defined in a test project for testing purposes https://github.com/OData/odata.net/blob/fce8fd9bf4942de679918ba3258d4b8dba7660f0/test/EndToEndTests/Tests/Client/Build.Desktop/Common/HttpWebRequestMessage.cs#L18. It implements `IODataRequestMessageAsync` and `IContainerProvider` interfaces.

File: src/Microsoft.OData.Edm/Schema/AmbiguousEntitySetBinding.cs (hunk 5 above)
Comment (ElizabethOkerio) [reply src/Microsoft.OData.Edm/Schema/AmbiguousEntitySetBinding.cs:46]: What of the property? The property does not need a doc string describing what it does? Also the method will not be removed per se...Maybe it will be converted to an internal or private method. But having it as obsolete is ok.

**Review by wachugamaina — DISMISSED** [review]

File: src/Microsoft.OData.Edm/Schema/BadEntitySet.cs (hunk 6 above)
Comment (gathogojr) [reply src/Microsoft.OData.Edm/Schema/BadEntitySet.cs:59]: The core implementations of `IEdmNavigationSource` are `EdmNavigationSource` and `CsdlSemanticsNavigationSource`

The following for implementations of `IEdmNavigationSource` are used in rare and special scenarios: `AmbiguousEntitySetBinding`, `AmbiguousSingletonBinding`, `BadEntitySet` and `UnknownEntitySet`. For these category of implementations, I have simply applied `/// <inheritdoc>`. That should suffice

File: src/Microsoft.OData.Edm/Schema/AmbiguousSingletonBinding.cs (hunk 7 above)
Comment (gathogojr) [reply src/Microsoft.OData.Edm/Schema/AmbiguousSingletonBinding.cs:28]: See this comment https://github.com/OData/odata.net/pull/2887#discussion_r1533480789

File: src/Microsoft.OData.Edm/Schema/AmbiguousEntitySetBinding.cs (hunk 5 above)
Comment (gathogojr) [reply src/Microsoft.OData.Edm/Schema/AmbiguousEntitySetBinding.cs:47]: REF: https://github.com/OData/odata.net/pull/2887#discussion_r1533480789

See this comment https://github.com/OData/odata.net/pull/2887#discussion_r1533480789

Comment (gathogojr) [reply src/Microsoft.OData.Edm/Schema/AmbiguousEntitySetBinding.cs:47]: REF: https://github.com/OData/odata.net/pull/2887#discussion_r1524302761

In the attempt to avoid repeating the statement `Type.AsElementType() as IEdmEntityType` that is repeated across `EdmNavigationSource`, `CsdlSemanticsNavigationSource`, `AmbiguousEntitySetBinding`, `AmbiguousSingletonBinding`, `BadEntitySet`, and `UnknownEntitySet`, I considered the following:
- Invoking the existing `EntityType` extension method from the the various implementations of `IEdmNavigationSource.EntityType` property. Considering that we have to add the `Obsolete` attribute to the extension method to signal that it'll cease being publicly available starting 9.0, this introduces warnings suggesting that the properties from where the deprecated method is called from should be also be marked with `Obsolete` attribute too. Suppressing those warnings with `pragma` statements makes the code look worse that it did before.
- Introducing a new internal extension method implemented in a similar manner as the deprecated `EntityType` extension method. This method would be called from the various implementations of `IEdmNavigationSource.EntityType` property. This option seems to introduce a problem to solve another for very little gain.

For the above reasons, I resolved to retain my initial implementation. The implementations are not subject to change with regularity so the angle of maintenance is moot (arguably)

**Review by habbes — APPROVED** [review]

**Review by ElizabethOkerio — APPROVED** [review]
